### DISCUSSION TITLE
How to prevent repository collaborators from triggering workflow

### DISCUSSION THREAD

**Discussion by northern-mindset** [community/community#27110] [Actions]
<p>According to <a href="https://help.github.com/en/github/automating-your-workflow-with-github-actions/authenticating-with-the-github_token" rel="nofollow noopener">https://help.github.com/en/github/automating-your-workflow-with-github-actions/authenticating-with-the-github_token</a>,</p>
<p><em>Anyone with write access to a repository can create, read, and use secrets.</em></p>
<p>Suppose there is a GitHub repository secret that contains a token for deploying to the staging (or even production) environment.</p>
<p>Any collaborator with write permissions can create a new GitHub workflow file, use the repository secret to deploys to staging/production environment in that workflow file, push the changes into any branch and thus trigger this workflow.</p>
<p>As a result, an arbitrary version of the code will be deployed.</p>
<p><strong>Is it possible to prevent this and allow triggering deployment only to authorized users?</strong></p>

**Comment by hypergeometric** [comment]
<p>I have the same thought.</p>
<p>If I want to use actions for deployments, I’d want to make sure that only master could be deployed to production. master can then be protected with branch protection to ensure that changes to workflow files must be approved by the relevent CODEOWNERS.</p>

**Comment by northern-mindset** [comment]
<p>According to <a href="https://github.community/t/question-on-actions-security/16597">https://github.community/t5/GitHub-Actions/Question-on-actions-security/m-p/35028/</a> it seems that in order to achieve desired security boundaries we have to use forks.</p>

**Comment by pedrocarloto** [comment]
<p><a class="mention" href="/anantakrishna">@anantakrishna</a> we have exactly the same problem. Have you figured out a solution without creating the overhead of a forked repository for every developer?</p>
<p>I have found this GitHub application: <a href="https://github.com/eladchen/protected-workflows#when-is-a-workflow-run-unauthorized" class="inline-onebox" rel="noopener nofollow ugc">GitHub - eladchen/protected-workflows: A Github application that cancels unauthorized workflow runs.</a><br>
that seems to solve the problem but I was trying to find a solution without recurring to third party applications.</p>
<p>I also found that BitBucket has this problem solved: <a href="https://bitbucket.org/blog/build-trust-in-your-deployment-workflow-with-deployment-permissions" class="inline-onebox" rel="noopener nofollow ugc">Build trust in your deployment workflow with deployment permissions - Bitbucket</a></p>
<p>Thank you.</p>

**Comment by northern-mindset** [comment]
<p>No, I have no news in this regard since then. So either use forks for collaborators (which is a pain, I should say) or move the deployment scripts (pipeline) to some other platform like Travis. Maybe having a separate repository on GitHub with the deployment actions might work, yet I have not tried it. Anyhow, not very convenient.</p>

**Comment by konradpabjan** [comment, marked as answer]
<p>Critical workflows and secrets can be put behind an environment that requires approval before running. The required approvers can be restricted to non-collaborators. See <a href="https://docs.github.com/en/actions/reference/environments" class="inline-onebox">Environments - GitHub Docs</a></p>
<p>This way any contributor could still queue the run, but there is an extra layer of protection and a mandatory review/approval would be required before the workflow would run.</p>

**Comment by northern-mindset** [comment]
<p>Indeed, this is something new! Thank you, <a class="mention" href="/konradpabjan">@konradpabjan</a>.</p>
<p>GitHub has released this feature just one month after me asking the question here:</p><aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://github.blog/wp-content/uploads/2019/01/cropped-github-favicon-512.png?fit=32%2C32" class="site-icon" width="32" height="32">
      <a href="https://github.blog/changelog/2020-12-15-github-actions-environments-environment-protection-rules-and-environment-secrets-beta/" target="_blank" rel="noopener nofollow ugc">The GitHub Blog</a>
  </header>
  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/443;"><img src="https://user-images.githubusercontent.com/6065071/181101731-20f1ec69-f9f1-499e-b598-11d32abb6d39.png" class="thumbnail" width="690" height="443" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/9/d/9d934875b8b01850d442f5f19e34d9a56905bb06_2_690x443.png, https://aws1.discourse-cdn.com/github/optimized/3X/9/d/9d934875b8b01850d442f5f19e34d9a56905bb06_2_1035x664.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/9/d/9d934875b8b01850d442f5f19e34d9a56905bb06.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/9/d/9d934875b8b01850d442f5f19e34d9a56905bb06_2_10x10.png"></div>

<h3><a href="https://github.blog/changelog/2020-12-15-github-actions-environments-environment-protection-rules-and-environment-secrets-beta/" target="_blank" rel="noopener nofollow ugc">GitHub Actions: Environments, environment protection rules and environment...</a></h3>

<p>GitHub Actions: Environments, environment protection rules and environment secrets (beta)</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>Here is a couple of follow-up blog posts about how to use this new feature:</p><aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://github.blog/wp-content/uploads/2019/01/cropped-github-favicon-512.png?fit=32%2C32" class="site-icon" width="32" height="32">
      <a href="https://github.blog/changelog/2021-02-17-github-actions-limit-which-branches-can-deploy-to-an-environment/" target="_blank" rel="noopener nofollow ugc">The GitHub Blog</a>
  </header>
  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:640/500;"><img src="https://user-images.githubusercontent.com/6065071/181101732-b8ae4bb7-350f-4d89-94f1-f9002e28d4fa.gif" class="thumbnail animated" width="640" height="500" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/3/a/3a11050962568d90eb0766a48788c3cadf23280b_2_10x10.png"></div>

<h3><a href="https://github.blog/changelog/2021-02-17-github-actions-limit-which-branches-can-deploy-to-an-environment/" target="_blank" rel="noopener nofollow ugc">GitHub Actions: Limit which branches can deploy to an environment | GitHub...</a></h3>

<p>GitHub Actions: Limit which branches can deploy to an environment</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://github.blog/wp-content/uploads/2019/01/cropped-github-favicon-512.png?fit=32%2C32" class="site-icon" width="32" height="32">
      <a href="https://github.blog/2021-04-13-implementing-least-privilege-for-secrets-in-github-actions/" target="_blank" rel="noopener nofollow ugc" title="06:05PM - 13 April 2021">The GitHub Blog – 13 Apr 21</a>
  </header>
  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/361;"><img src="https://github.blog/wp-content/uploads/2019/03/product-social.png?fit=1201%2C630" class="thumbnail" width="690" height="361"></div>

<h3><a href="https://github.blog/2021-04-13-implementing-least-privilege-for-secrets-in-github-actions/" target="_blank" rel="noopener nofollow ugc">Implementing least privilege for secrets in GitHub Actions | The GitHub Blog</a></h3>

<p>GitHub Actions provide several features to help your organization effectively implement a secret management strategy based on least privilege.</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>Now I’m eager to adopt this feature in my projects!</p>
